io/src/main/scala/sbt/internal/io/EventMonitor.scala

```diff
@@ -0,0 +1,194 @@
+package sbt.internal.io
+
+import java.nio.file.{ ClosedWatchServiceException, Files, Path, WatchKey }
+import java.util.concurrent.ArrayBlockingQueue
+import java.util.concurrent.atomic.{ AtomicBoolean, AtomicInteger }
+
+import sbt.io.WatchService
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+
+private[sbt] sealed trait EventMonitor extends AutoCloseable {
+
+  /** A snapshot of the WatchState that includes the number of build triggers and watch sources. */
+  def state(): WatchState
+
+  /** Block indefinitely until the monitor receives a file event or the user stops the watch. */
+  def watch(): Boolean
```

Comment:
I wonder if there might be a better blocky name, like `poll()` or `awaitEvent()`.